### ISSUE TITLE
Training, testing and validating (Graphsage example)

### ISSUE THREAD

**Issue by zuenko** [dmlc/dgl#455]
Hello!
## ❓ Questions and Help
I don't get how is it possible that we are learning on the whole graph *[line of code](https://github.com/dmlc/dgl/blob/master/examples/pytorch/graphsage/graphsage.py#L213), then we use some "mask" to compute the loss *[line of code](https://github.com/dmlc/dgl/blob/master/examples/pytorch/graphsage/graphsage.py#L214). 
Another thing is about the loss in the original implementation: *[line of code](https://github.com/williamleif/GraphSAGE/blob/master/graphsage/supervised_models.py#L102), as you can see they use an l2 loss in each aggregator and only then softmax or sigmoid. 

Is it just an example or implementation with a mistake? 

P.s. 
> Before proceeding, please note that we recommend
> using our discussion forum (https://discuss.dgl.ai) for
> general questions. As a result, this issue will
> likely be CLOSED shortly.

Recommend using the forum, constrain new users :(

**Comment by jermainewang** [comment]
@BarclayII Could you comment on this?

> P.s.
> 
> > Before proceeding, please note that we recommend
> > using our discussion forum (https://discuss.dgl.ai) for
> > general questions. As a result, this issue will
> > likely be CLOSED shortly.
> 
> Recommend using the forum, constrain new users :(

We recommend use forum so that the issue list on github could be more relevant to features/bugs/roadmap. Otherwise, these important issues will quickly sink to the bottom. It also gives an impression that this project has "a lot of issues". On the other hand, using forum gives you many benefit: (1) very good search bar to find relevant questions; (2) very good reply systems that support rich features; (3) waterfall view so you only need to scroll down; (4) other forum features such as finding topic by certain users. We also post news to forum so people could discuss (which is apparently not an *issue*). It is also how many other projects manage communities (e.g. [pytorch](https://discuss.pytorch.org/), [mxnet](https://discuss.mxnet.io/)). Registering should be minimal (I remember only verifying email is required).  Let me know if you have any problem using the forum.

**Comment by zuenko** [comment]
@jermainewang "Sorry, new users can only put 2 links in a post." :(

**Comment by jermainewang** [comment]
I've changed that to 10. Thanks for the suggestion.

**Comment by BarclayII** [comment]
A clarification: this is a simplified implementation of GraphSAGE, in the sense that we are not sampling the neighborhood as the authors did.

> I don't get how is it possible that we are learning on the whole graph *[line of code](https://github.com/dmlc/dgl/blob/master/examples/pytorch/graphsage/graphsage.py#L213), then we use some "mask" to compute the loss *[line of code](https://github.com/dmlc/dgl/blob/master/examples/pytorch/graphsage/graphsage.py#L214).

Recall that GraphSAGE classifies nodes in a given graph.  So what we did is simply
* Predict the class of *every single* node, regardless of which data split (training, validation, or test) the node is in.  ([L213]((https://github.com/dmlc/dgl/blob/master/examples/pytorch/graphsage/graphsage.py#L213)))
* Afterwards, we only inspect and compute loss from the nodes in the training set.  So the model is not looking at either validation or test set during training.  ([L214](https://github.com/dmlc/dgl/blob/master/examples/pytorch/graphsage/graphsage.py#L214))

> Another thing is about the loss in the original implementation: *[line of code](https://github.com/williamleif/GraphSAGE/blob/master/graphsage/supervised_models.py#L102), as you can see they use an l2 loss in each aggregator and only then softmax or sigmoid.
> 
> Is it just an example or implementation with a mistake?

The L2 loss in the original Tensorflow implementation is actually the L2 regularization on model weights, which is handled by the `weight_decay` argument in the optimizer ([L204](https://github.com/dmlc/dgl/blob/master/examples/pytorch/graphsage/graphsage.py#L204)).  Thus we only need to `backward` the cross entropy loss.

Let us know if you have further questions.
